### ISSUE TITLE
Emailing ActionText rich_text field shows attachment links to example.org

### ISSUE THREAD

**Issue by davetron5000** [rails/rails#35578]
### Steps to reproduce

Vanilla setup:

```ruby
class SupportRequest < ApplicationRecord
  belongs_to :order, optional: true

  has_rich_text :response
end
```

Capturing and rendering this in a web view works great.

I want to email the rich text such that links to inlined images work:

```ruby
class SupportRequestResponseMailer < ApplicationMailer
  def respond(support_request)
    @support_request = support_request
    mail(to: support_request.order.email,
         from: "support@example.com",
         content_type: "text/html",
         body: support_request.response.to_s)
  end
end
```

When I do this, the email's `img` tags link to `example.org`.  I *cannot* figure out what is happening.

`config/environments/development.rb`:

```ruby
Rails.application.configure do
  # Settings specified here will take precedence over those in config/application.rb.

  # In the development environment your application's code is reloaded on
  # every request. This slows down response time but is perfect for development
  # since you don't have to restart the web server when you make code changes.
  config.cache_classes = false

  # Do not eager load code on boot.
  config.eager_load = false

  # Show full error reports.
  config.consider_all_requests_local = true

  # Enable/disable caching. By default caching is disabled.
  if Rails.root.join('tmp/caching-dev.txt').exist?
    config.action_controller.perform_caching = true

    config.cache_store = :memory_store
    config.public_file_server.headers = {
      'Cache-Control' => "public, max-age=#{2.days.seconds.to_i}"
    }
  else
    config.action_controller.perform_caching = false

    config.cache_store = :null_store
  end

  # Don't care if the mailer can't send.
  config.action_mailer.raise_delivery_errors = false
#START_HIGHLIGHT

  # Don't actually send emails
  config.action_mailer.delivery_method = :smtp
  config.action_mailer.perform_deliveries = true
  config.action_mailer.raise_delivery_errors = true
  config.action_mailer.smtp_settings = { :address => "localhost", :port => 1025 }
  config.action_mailer.default_url_options = {:host => "localhost", :port => '3000', :protocol => "http"}
  config.action_mailer.preview_path = "#{Rails.root}/tmp/mailers/previews"
  Rails.application.routes.default_url_options[:host] = 'localhost:3000'
  #
  # Alternate configuration example, using gmail:
  #   config.action_mailer.delivery_method = :smtp
  #   config.action_mailer.smtp_settings = {
  #     address:        "smtp.gmail.com",
  #     port:           587, 
  #     domain:         "domain.of.sender.net",
  #     authentication: "plain",
  #     user_name:      "dave",
  #     password:       "secret",
  #     enable_starttls_auto: true
  #   } 
#END_HIGHLIGHT

  config.action_mailer.perform_caching = false

  # Print deprecation notices to the Rails logger.
  config.active_support.deprecation = :log

  # Raise an error on page load if there are pending migrations.
  config.active_record.migration_error = :page_load

  # Debug mode disables concatenation and preprocessing of assets.
  # This option may cause significant delays in view rendering with a large
  # number of complex assets.
  config.assets.debug = true

  # Suppress logger output for asset requests.
  config.assets.quiet = true

  # Raises error for missing translations
  # config.action_view.raise_on_missing_translations = true

  # Use an evented file watcher to asynchronously detect changes in source code,
  # routes, locales, etc. This feature depends on the listen gem.
  # config.file_watcher = ActiveSupport::EventedFileUpdateChecker
  config.active_storage.service = :local
end
```
### Expected behavior

I was hoping that the `img` src in the email content would be to `localhost:3000` (my dev server) and not `example.org`

### Actual behavior

`img` src is to `example.org`.  Also does this in `bin/rails c`

### System configuration
**Rails version**: 6.0.0.beta2

**Ruby version**: `ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-darwin17]`

**Comment by prathamesh-sonpatki** [comment]
Can you upload a sample application on Github that reproduces the issue? Thanks.

**Comment by davetron5000** [comment]
@prathamesh-sonpatki Yup: https://github.com/davetron5000/depot_u

The `README` has instructions on how to use the app to reveal the problem.

Thanks!

**Comment by markvanlan** [comment]
I am having a very similar issue. ActionText is not using the same asset_host to show images where they are rendered, as where the image is uploaded. 

In my example, the image was uploaded from `engage.lvh.ne:3000`, but the image is being rendered on the domain `i91.lvh.me:3000`.

I tried setting `config.action_controller.asset_host = "engage.lvh.me:3000"`, but that did not change anything. 

Here is a screenshot of the html being generated on `i91.lvh.me:3000`
![issue](https://user-images.githubusercontent.com/16214023/57540330-1cb8f600-7312-11e9-883a-0a672c42f112.png)

**Comment by jeffschuil** [comment]
@davetron5000 You might want to look at your `app/views/active_storage/blobs/_blob.html.haml` which is used to generate the attachment HTML (i.e. image tags).

https://github.com/davetron5000/depot_u/blob/efe36dec8007648b81bf025854c973763487ae9a/app/views/active_storage/blobs/_blob.html.erb#L3

When rendering from a mailer, I tried wrapping the `blob.representation` in `url_for` and it appeared to work.

`image_tag url_for(blob.representation(resize_to_limit: local_assigns[:in_gallery] ? [ 800, 600 ] : [ 1024, 768 ]))`

Some confirmation that this is the right thing to do would be good, though.

I also tried using the `service_url` for the ActiveStorage blob, but those URLs are meant to be temporary, which means the email eventually had broken images.

**Comment by jeffschuil** [comment]
Similarly, I experimented with using `rails_blob_url`, passing a host, which helped with the links.

`rails_blob_url(blob, { host: 'example.com' })`

In my case, I reused the host set in:

`Rails.application.config.action_mailer.default_url_options`

**Comment by jeffblake** [comment]
Since ActionText uses `ApplicationController.renderer`, I would expect that the settings in the `application_controller_renderer.rb` initializer would be used, but they are not.

For example I have 

```
ActiveSupport::Reloader.to_prepare do
  ApplicationController.renderer.defaults.merge!(
    http_host: ENV.fetch('URL_HOST'),
    https: false
  )
end
```
But this has no effect. The only thing that works is to do this hack...

```
ActionController::Renderer.const_set('DEFAULTS', {
  http_host: ENV.fetch('URL_HOST'),
  https: false,
  method: "get",
  script_name: "",
  input: ""
}.freeze
)
```

**Comment by adammiribyan** [comment]
👍

**Comment by palytoxin** [comment]
I have same issue, I use jsonapi_resource to fetch actiontext rich text, but the blob url is all from example.org

**Comment by rigbooks** [comment]
The only sane way I found to get this to work correctly (and I tried several ways) was to override default_url_options in ActionController

```
  def default_url_options(options = {})
    {
      host: Rails.application.credentials.dig(:app, :host), 
      port: Rails.application.credentials.dig(:app, :port), 
      protocol: Rails.application.credentials.dig(:app, :protocol)
    }.merge(options)
  end
```

Ideally, ActionText would function the way ActionMailer does and assign host options when rendering.

**Comment by weilandia** [comment]
Experiencing the same issue. Resolved with:

```
<%= image_tag Rails.application.routes.url_helpers.rails_blob_url(blob) %>
```

and

```
<a href="<%= Rails.application.routes.url_helpers.rails_blob_url(blob, disposition: "attachment") %>">
```

in _blob.html.erb

**Comment by agrobbin** [comment]
I solved this in a slightly different way, following a similar pattern that Action Text itself uses to set its `renderer`:

```ruby
ActiveSupport.on_load(:action_mailer) do
  helper ActionText::Engine.helpers

  around_action do |_mailer, action|
    original_renderer = ActionText::Content.renderer
    ActionText::Content.renderer = self

    action.call

    ActionText::Content.renderer = original_renderer
  end
end
```

This uses Action Mailer to render Action Text while in the context of Action Mailer itself, resetting the `renderer` back to its original state once finished processing the mailer action (which should handle any cases of using `deliver_now` within the context of a controller action).

This should make the `ActionController::Renderer::DEFAULTS` patches no longer necessary.

**Comment by pinzonjulian** [comment]
Same thing happens to me and @agrobbin 's  solution works just fine! Any chance of getting this issue reopened?

Steps to reproduce:
1. Create new rails app
2. Generate a model with a rich text field
    `rails g model post body:rich_text`
3. Generate a new mailer with an example delivery method
    `rails g mailer example send_post`
4. Create a post with rich text in it that has an image attached to it (use a scaffold for this if you'd like)
5. Modify the `send_post` method to get any `Post`. 
6. Modify the `send_post.html.erb` template to use the post's rich_text body: `@post.body`
7. To view the sent email, add the `letter_opener` gem
8. Send email using the rails console
9. See email in your browser
10. Email's image won't be rendered.

**Comment by vfonic** [comment]
I was hit by this as well yesterday, sending an email to almost 600 customers.

This is a bug. There's no documented way on how to fix this so users are finding workarounds.

It would help a lot if this issue could be reopened and then we can look at what would be the best way to fix this.

Thanks!

**Comment by vfonic** [comment]
This seems to have been fixed in rails 6.1.0.rc1. I updated my app to use that rails version and it's working without me making any other changes.

This is the commit that most likely fixed it: https://github.com/rails/rails/commit/614e813161ad8f6e13ad19aedd577acdd976d9d6
